src/server/cluster/cluster_slot_migration.cc

```diff
@@ -82,4 +100,73 @@ ClusterSlotMigration::Info ClusterSlotMigration::GetInfo() const {
   return {ctx.host, ctx.port, state_};
 }
 
+void ClusterSlotMigration::MainMigrationFb() {
+  VLOG(1) << "Main migration fiber started";
+
+  state_ = ClusterSlotMigration::C_FULL_SYNC;
+
+  // TODO add reconnection code
+  if (auto ec = InitiateDflyFullSync(); ec) {
+    LOG(WARNING) << "Error syncing with " << server().Description() << " " << ec << " "
+                 << ec.message();
+  }
+}
+
+std::error_code ClusterSlotMigration::InitiateDflyFullSync() {
+  shard_flows_.resize(source_shards_num_);
+  for (unsigned i = 0; i < source_shards_num_; ++i) {
+    shard_flows_[i].reset(new ClusterShardMigration(server(), i));
+  }
+
+  // Blocked on until all flows got full sync cut.
+  BlockingCounter sync_block{source_shards_num_};
+
+  // Switch to new error handler that closes flow sockets.
+  auto err_handler = [this, sync_block](const auto& ge) mutable {
+    // Unblock this function.
+    sync_block.Cancel();
+
+    // Make sure the flows are not in a state transition
+    lock_guard lk{flows_op_mu_};
+
+    // Unblock all sockets.
+    DefaultErrorHandler(ge);
+    for (auto& flow : shard_flows_)
+      flow->Cancel();
+  };
+  RETURN_ON_ERR(cntx_.SwitchErrorHandler(std::move(err_handler)));
+
+  CHECK(service_.SwitchState(GlobalState::ACTIVE, GlobalState::LOADING).first ==
+        GlobalState::LOADING);
+
+  absl::Cleanup cleanup = [this]() {
+    // We do the following operations regardless of outcome.
+    service_.SwitchState(GlobalState::LOADING, GlobalState::ACTIVE);
+  };
+
+  std::atomic_uint32_t synced_shards = 0;
+  auto partition = Partition(source_shards_num_);
+  auto shard_cb = [&](unsigned index, auto*) {
+    for (auto id : partition[index]) {
+      auto ec = shard_flows_[id]->StartSyncFlow(sync_block, &cntx_);
+      if (!ec) {
+        ++synced_shards;
+      } else {
+        cntx_.ReportError(ec);
+      }
+    }
+  };
+  // Lock to prevent the error handler from running instantly
+  // while the flows are in a mixed state.
+  lock_guard lk{flows_op_mu_};
+  shard_set->pool()->AwaitFiberOnAll(std::move(shard_cb));
+
+  VLOG(1) << synced_shards << " from " << source_shards_num_ << " shards is synchronized";
+  if (synced_shards != source_shards_num_) {
+    cntx_.ReportError(std::make_error_code(errc::state_not_recoverable), "full sync is failed");
+  }
+
+  return cntx_.GetError();
```

Comment:
This is a full copy of our replica. Didn't we really evaluate the choice of modifying/extending it to support the cluster protocol? There are literally only a few lines that differ (apologies to not pointing this out on the previous PR)

Everything we do with replication has to be duplicated here

Reply:
I think it should be very different from the replica code, and made some comments on this PR that shows the difference. As I see it the simillarity between cluster migration and replica is that we need to have migration per shard as we have replication per shard and the executer which will read parse the journal and will execute it. We will have also the main fiber that flow from start if we need to reconnect.

Reply:
> it should be very different from the replica code

But currently it is very similar 😆 At least the high level logic or the general concept